poc/find_wr_params.sage

```diff
@@ -60,252 +81,426 @@ def flp_sound(valid: PineValid):
-    complete the parameters, we run this on the output of `search()`, make an
-    initial guess of `eta`, and manually tune until the desired ZK is reached.
+    Search for the optimal wraparound check parameters for each set of user
+    parameters in `user_params`. The search logic works as follows:
+    - Find all pairs of `num_wr_checks` and `num_wr_successes` that can satisfy
+      soundness error. We do so by finding the `num_wr_successes`, for each
+      `num_wr_checks` from `target_soundness_bits` up to `max_num_wr_checks`,
+      with a step of `num_wr_checks_step`. For each `num_wr_checks`, the search
+      logic is handled in `search_num_wr_successes`.
+    - Then for each set of parameters in `user_params`, go through the pairs of
+      `num_wr_checks` and `num_wr_successes`, and find a viable `alpha`. For
+      each possible combination of `num_wr_checks`, `num_wr_successes`, and
+      `alpha`, compute the overhead in terms of the number of field elements
+      needed to encode the wraparound check results. We pick the combination
+      that gives the lowest overhead.
     '''
+    # The search range for `alpha`.
+    # We pick this maximum alpha, because it ensures a sufficiently low ZK error
+    # for all user parameters.
+    max_alpha = 10
+    min_alpha = compute_min_alpha()
+
+    # Find all pairs of `num_wr_checks` and `num_wr_successes`.
+    num_wr_checks_and_successes = []
+    for num_wr_checks in range(target_soundness_bits,
+                               max_num_wr_checks + 1,
+                               num_wr_checks_step):
+        num_wr_successes = search_num_wr_successes(target_soundness_bits,
+                                                   num_wr_checks)
+        num_wr_checks_and_successes.append((num_wr_checks, num_wr_successes))
+
+    # Collect derived wraparound check parameters and user parameters that
+    # satisfy the target ZK error.
+    wr_and_user_params = []
+    for (l2_norm_bound, num_frac_bits, _, field) in user_params:
+        # Record the combination of `num_wr_checks`, `num_wr_successes`, and
+        # `alpha` that gives the lowest overhead in terms of the number of
+        # field elements to encode range-checked, wraparound check results.
+        best_num_wr_checks = None
+        best_num_wr_successes = None
+        best_alpha = None
+        best_overhead = None
+        for (num_wr_checks, num_wr_successes) in num_wr_checks_and_successes:
+            # Search for `alpha` in `[min_alpha, max_alpha]`, given
+            # `num_wr_checks` and `num_wr_successes`.
+            curr_min_alpha = min_alpha
+            curr_max_alpha = max_alpha
+            alpha = None
+            # Do a binary search between `curr_min_alpha` and `curr_max_alpha`,
+            # and stop when `curr_min_alpha` is at least `curr_max_alpha`, or
+            # the difference between `curr_min_alpha` and `curr_max_alpha` is
+            # < 0.1.
```

Comment:
There are two binary searches in this program. Consider refactoring so that they use the same function.